src/utils/memory_pool.hpp

```diff
@@ -0,0 +1,274 @@
+
+  weak_t Get();
+
+  void PrintStatistics()
+  {
+    std::cout << available_.size() << " unused objects." << std::endl;
+    std::cout << inuse_.size() << " used objects." << std::endl;
+  }
+
+private:
+  bool IsValid(const weak_t &in) const
+  {
+    return inuse_.count(in.key_);
+  }
+
+  void ReferenceCountedFree(const weak_t &in)
+  {
+    if (!IsValid(in))
+      return;
+    auto &pair = inuse_[in.key_];
+    --pair.second;
+    if (pair.second <= 0)
+    {
+      available_.push(pair.first);
+      inuse_.erase(in.key_);
+    }
+  }
+
+  void Free(const weak_t &in)
+  {
+    if (!IsValid(in))
+      return;
+    available_.push(inuse_[in.key_].first);
+    inuse_.erase(in.key_);
+  }
+
+  void AddCount(const weak_t &in)
+  {
+    if (!IsValid(in))
+      throw 1;
+    ++inuse_[in.key_].second;
+  }
+};
+
+// Lightly wraps a view of type T and holds a key that
+// that can be used to query if its storage
+// is in use or if it has been freed and allows
+// freeing of the storage.
+template <class T, class KEY_T>
+struct Pool<T, KEY_T>::weak_t : public T
+{
+  friend class Pool;
+
+protected:
+  template <class... Ts>
+  KOKKOS_IMPL_HOST_FUNCTION static weak_t make(int key, Ts &&...args)
+  {
+    weak_t out(args...);
+    out.key_ = key;
+    return out;
+  }
+
+public:
+  template <class... Ts>
+  KOKKOS_IMPL_HOST_FUNCTION
+  weak_t(Ts &&...args) : T(std::forward<Ts>(args)...), key_(default_key_){};
+
+  KOKKOS_IMPL_HOST_FUNCTION
+  inline void Free() { (*pool_).Free(*this); }
+
+  KOKKOS_IMPL_HOST_FUNCTION
+  inline bool IsValid()
+  {
+    if (key_ == default_key_ || pool_ == nullptr)
+      return false;
+    return (*pool_).IsValid(*this);
+  }
+
+  KOKKOS_IMPL_HOST_FUNCTION
+  inline KEY_T GetKey() const { return key_; }
+
+  KOKKOS_DEFAULTED_FUNCTION
+  ~weak_t() = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t() = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t(const weak_t &) = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t(weak_t &&) = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t &operator=(const weak_t &) = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t &operator=(weak_t &&) = default;
+  
+  // Allow this to point at an unmanaged object of type T
+  KOKKOS_FUNCTION 
+  weak_t &operator=(const T& in) {
+    T::operator=(in);
+    return *this;
+  }
+
+protected:
+  KEY_T key_ = 0;
+  Pool *pool_ = nullptr;
+};
+
+// Reference counted version of pool member that has ownership over a resource
+// and sends it back to the pool when its destructor is called and it is the
+// last owner that holds that resource. Cannot be on device
+template <class T, class KEY_T>
+class Pool<T, KEY_T>::owner_t : public Pool<T, KEY_T>::weak_t
+{
+public:
+  KOKKOS_DEFAULTED_FUNCTION
+  owner_t() = default;
+
+  KOKKOS_FUNCTION
+  ~owner_t() noexcept
+  {
+#ifndef __CUDA_ARCH__ // host code
+    if (weak_t::pool_ != nullptr)
+      (*weak_t::pool_).ReferenceCountedFree(*this);
+#endif
```

Comment:
I agree that I would like to get rid of this, but I spent a little while searching around and couldn't figure out a way with Kokkos. If anyone has an idea, I would certainly be open to such a change.

Reply:
we might be able to query the memory space the view is on. I've done this in spiner:
https://github.com/lanl/spiner/blob/main/spiner/databox.hpp#L287

Reply:
Perhaps this should be turned into an issue. It would be good to change, but figuring out a solution is not high priority right now from my perspective.